org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ProjectUtils.java

```diff
@@ -83,4 +95,180 @@ public static String getJavaSourceLevel(IProject project) {
+		List<IClasspathEntry> newEntries = new ArrayList<>();
+		boolean found = false;
+		for (IClasspathEntry entry : existingEntries) {
+			if (entry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+				if (entry.getPath().equals(sourcePath)) {
+					found = true;
+				} else {
+					newEntries.add(removeFilters(entry, sourcePath));
+				}
+			} else {
+				newEntries.add(entry);
+			}
+		}
+
+		if (found) {
+			project.setRawClasspath(newEntries.toArray(new IClasspathEntry[0]), project.getOutputLocation(), null);
+		} else {
+			throw new CoreException(
+					new Status(IStatus.ERROR, JavaLanguageServerPlugin.PLUGIN_ID, Messages.format("The folder ''{0}'' doesn''t belong to any project''s source path. No need of the removal operation.", sourcePath.toOSString())));
+		}
+	}
+
+	public static IPath[] listSourcePaths(IJavaProject project) throws JavaModelException {
+		List<IPath> result = new ArrayList<>();
+		for (IClasspathEntry entry : project.getRawClasspath()) {
+			if (entry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+				result.add(entry.getPath());
+			}
+		}
+
+		return result.toArray(new IPath[0]);
+	}
+
+	public static boolean isOnSourcePath(IPath sourcePath, IJavaProject project) throws JavaModelException {
+		for (IClasspathEntry entry : project.getRawClasspath()) {
+			if (entry.getEntryKind() == IClasspathEntry.CPE_SOURCE && entry.getPath().equals(sourcePath)) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public static IPath findBelongedWorkspaceRoot(IPath filePath) {
+		PreferenceManager manager = JavaLanguageServerPlugin.getPreferencesManager();
+		Collection<IPath> rootPaths = manager.getPreferences().getRootPaths();
+		if (rootPaths != null) {
+			for (IPath rootPath : rootPaths) {
+				if (rootPath.isPrefixOf(filePath)) {
+					return rootPath;
+				}
+			}
+		}
+
+		return null;
+	}
+
+	public static String getWorkspaceInvisibleProjectName(IPath workspacePath) {
+		String fileName = workspacePath.toFile().getName();
+		String projectName = fileName + "_" + Integer.toHexString(workspacePath.toPortableString().hashCode());
+		return projectName;
+	}
+
+	public static boolean isVisibleProject(IProject project) {
+		PreferenceManager manager = JavaLanguageServerPlugin.getPreferencesManager();
+		Collection<IPath> rootPaths = manager.getPreferences().getRootPaths();
+		if (rootPaths == null) {
+			return false;
+		}
+
+		return ResourceUtils.isContainedIn(project.getLocation(), rootPaths);
+	}
+
+	public static List<IProject> getVisibleProjects(IPath workspaceRoot) {
+		List<IProject> projects = new ArrayList<>();
+		for (IProject project : ResourcesPlugin.getWorkspace().getRoot().getProjects()) {
+			if (project.exists() && isVisibleProject(project) && workspaceRoot.isPrefixOf(project.getLocation())) {
+				projects.add(project);
+			}
+		}
+
+		return projects;
+	}
+
+	public static IProject createInvisibleProjectIfNotExist(IPath workspaceRoot) throws OperationCanceledException, CoreException {
```

Comment:
Share the same InvisibleProject concept for the workspace without project descriptor. But the code doesn't have dependency because this is just a common utility method.